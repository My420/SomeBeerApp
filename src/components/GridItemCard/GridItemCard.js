import React from 'react';
import PropTypes from 'prop-types';
import styles from './GridItemCard.module.scss';
import GridCardControl from './GridCardControl/GridCardControl';
import GridCardLink from './GridCardLink/GridCardLink';

class GridItemCard extends React.Component {
  state = { isDescriptionOpen: false };

  showDescription = () => {
    this.setState({ isDescriptionOpen: true });
  };

  hideDescription = () => {
    this.setState({ isDescriptionOpen: false });
  };

  toggleDescription = () => {
    const { isDescriptionOpen } = this.state;
    this.setState({ isDescriptionOpen: !isDescriptionOpen });
  };

  addEndListener = (node, done) =>
    node.addEventListener('transitionend', done, false);

  render() {
    const { data } = this.props;
    const { isDescriptionOpen } = this.state;

    return (
      <div className={styles.card}>
        <GridCardLink
          data={data}
          isDescriptionOpen={isDescriptionOpen}
          onCursorEnter={this.showDescription}
          onCursorLeave={this.hideDescription}
        />
        <GridCardControl
          data={data}
          isDescriptionOpen={isDescriptionOpen}
          onToggleButtonClick={this.toggleDescription}
        />
      </div>
    );
  }
}

GridItemCard.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.object.isRequired
};

export default GridItemCard;
