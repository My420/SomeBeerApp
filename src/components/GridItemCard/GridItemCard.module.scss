@import '../../sass/variables.scss';

.card {
  font-size: 18px;
  line-height: 24px;
  border: 2px solid $menu-bar-color;
}

.card:hover {
  box-shadow: 0 0 5px 1px $menu-color;
}
