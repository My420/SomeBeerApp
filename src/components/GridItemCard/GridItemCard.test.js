import React from 'react';
import { shallow } from 'enzyme';
import GridItemCard from './GridItemCard';

describe('test <GridItemCard/> component', () => {
  const data = { id: 1 };

  describe('test component with default state', () => {
    const expectedState = {
      isDescriptionOpen: false
    };

    const wrapper = shallow(<GridItemCard data={data} />);

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should render component with correct state', () => {
      expect(wrapper.state()).toEqual(expectedState);
    });

    test('should render GridCardLink component with correct prop data', () => {
      const component = wrapper.find('GridCardLink');
      expect(component.prop('data')).toEqual(data);
      expect(component.prop('isDescriptionOpen')).toBe(
        expectedState.isDescriptionOpen
      );
    });

    test('should render GridCardControl component with correct prop data', () => {
      const component = wrapper.find('GridCardControl');
      expect(component.prop('data')).toEqual(data);
      expect(component.prop('isDescriptionOpen')).toBe(
        expectedState.isDescriptionOpen
      );
    });
  });

  describe('test component when state change', () => {
    const newState = {
      isDescriptionOpen: true
    };

    const wrapper = shallow(<GridItemCard data={data} />);

    wrapper.setState({ ...newState });

    test('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    test('should render component with correct state', () => {
      expect(wrapper.state()).toEqual(newState);
    });

    test('should render GridCardLink component with correct prop data', () => {
      const component = wrapper.find('GridCardLink');
      expect(component.prop('data')).toEqual(data);
      expect(component.prop('isDescriptionOpen')).toBe(
        newState.isDescriptionOpen
      );
    });

    test('should render GridCardControl component with correct prop data', () => {
      const component = wrapper.find('GridCardControl');
      expect(component.prop('data')).toEqual(data);
      expect(component.prop('isDescriptionOpen')).toBe(
        newState.isDescriptionOpen
      );
    });
  });

  describe('test components methods', () => {
    test('showDescription method should work correct', () => {
      const wrapper = shallow(<GridItemCard data={data} />);
      const instance = wrapper.instance();
      expect(wrapper.state().isDescriptionOpen).toBeFalsy();
      instance.showDescription();
      expect(wrapper.state().isDescriptionOpen).toBeTruthy();
    });

    test('hideDescription method should work correct', () => {
      const wrapper = shallow(<GridItemCard data={data} />);
      wrapper.setState({ isDescriptionOpen: true });
      const instance = wrapper.instance();
      expect(wrapper.state().isDescriptionOpen).toBeTruthy();
      instance.hideDescription();
      expect(wrapper.state().isDescriptionOpen).toBeFalsy();
    });

    test('toggleDescription method should work correct', () => {
      const wrapper = shallow(<GridItemCard data={data} />);
      const instance = wrapper.instance();
      expect(wrapper.state().isDescriptionOpen).toBeFalsy();
      instance.toggleDescription();
      expect(wrapper.state().isDescriptionOpen).toBeTruthy();
      instance.toggleDescription();
      expect(wrapper.state().isDescriptionOpen).toBeFalsy();
    });
  });
});
