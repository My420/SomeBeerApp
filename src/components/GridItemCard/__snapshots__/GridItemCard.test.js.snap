// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <GridItemCard/> component test component when state change should match snapshot 1`] = `
<div
  className="card"
>
  <GridCardLink
    data={
      Object {
        "id": 1,
      }
    }
    isDescriptionOpen={true}
    onCursorEnter={[Function]}
    onCursorLeave={[Function]}
  />
  <GridCardControl
    data={
      Object {
        "id": 1,
      }
    }
    isDescriptionOpen={true}
    onToggleButtonClick={[Function]}
  />
</div>
`;

exports[`test <GridItemCard/> component test component with default state should match snapshot 1`] = `
<div
  className="card"
>
  <GridCardLink
    data={
      Object {
        "id": 1,
      }
    }
    isDescriptionOpen={false}
    onCursorEnter={[Function]}
    onCursorLeave={[Function]}
  />
  <GridCardControl
    data={
      Object {
        "id": 1,
      }
    }
    isDescriptionOpen={false}
    onToggleButtonClick={[Function]}
  />
</div>
`;
